File: packages/study/agent-go-blocks/pkg/agent/memory.go
package agent

import "context"

type ChatMemoryBuffer struct {
	history []ChatMessage
}

func NewChatMemoryBuffer(initialHistory []ChatMessage) *ChatMemoryBuffer {
	return &ChatMemoryBuffer{history: initialHistory}
}

func (m *ChatMemoryBuffer) BeforeLLMCall(ctx context.Context, messages []ChatMessage) ([]ChatMessage, error) {
	return append(m.history, messages...), nil
}

func (m *ChatMemoryBuffer) AfterLLMCall(ctx context.Context, userMessage, assistantResponse string) error {
	m.history = append(m.history, ChatMessage{Role: RoleUser, Content: userMessage})
	m.history = append(m.history, ChatMessage{Role: RoleModel, Content: assistantResponse})
	return nil
}

func (m *ChatMemoryBuffer) Reset() {
	m.history = nil
}

File: packages/study/agent-rs-blocks/src/agent/memory.rs
use crate::agent::types::{AgentComponent, ChatMessage, MessageRole};
use async_trait::async_trait;

pub struct ChatMemoryBuffer {
    history: Vec<ChatMessage>,
}

impl ChatMemoryBuffer {
    pub fn new(initial_history: Vec<ChatMessage>) -> Self {
        Self { history: initial_history }
    }

    pub fn reset(&mut self) {
        self.history.clear();
    }
}

#[async_trait]
impl AgentComponent for ChatMemoryBuffer {
    async fn before_llm_call(&self, messages: Vec<ChatMessage>) -> Vec<ChatMessage> {
        let mut new_messages = self.history.clone();
        new_messages.extend(messages);
        new_messages
    }

    async fn after_llm_call(&mut self, user_message: &str, assistant_response: &str) {
        self.history.push(ChatMessage {
            role: MessageRole::User,
            content: user_message.to_string(),
        });
        self.history.push(ChatMessage {
            role: MessageRole::Model,
            content: assistant_response.to_string(),
        });
    }
}
